src/Query/AbstractCryptParser.php

```diff
@@ -0,0 +1,22 @@
+<?php
+
+namespace DoctrineExtensions\Query;
+
+use Doctrine\ORM\Query\AST\Functions\FunctionNode;
+use Doctrine\ORM\Query\Lexer;
+use Doctrine\ORM\Query\TokenType;
+
+abstract class AbstractCryptParser extends FunctionNode
```

Comment:
What do we need all those new abstract classes for?

Reply:
There was a lot of duplicated code, often copy paste and the property name was not correct (it was called date but did not store a date related node). Eliminate this duplication and introduce a more generic property name.

Reply:
I understand that abstracting duplicated code and renaming properties might be desirable in general, but this way you're introducing a lot of unrelated changes which might render the code review difficult.

Reply:
If you tell me that the separate PR will be merged and not get stuck in the queue of waiting PRs, I'll be happy to split this PR into 4 parts (which is a draft anyway, so I don't expect a review)

Reply:
It's @beberlei's repository, so I can't make any promises on what will be merged and what not. But I would try to support you and provide reviews, so that it's easier for @beberlei to hit the merge button eventually. 🙂

Reply:
> should there be a CI php8.3 test? yes / no

Yes, but PHP 8.3 is completely irrelevant for ORM 3 compatibility.

> should cs-fixer 3.x upgrade? yes / no

Probably, but again: Not relevant for ORM 3.

> should there be consistency in the method order. This can be guaranteed by cs-fixer, see `ordered_class_elements` fixer: properties, getSql method, parse method (I like parts that do the same thing to look consistent)? yes / no

I have no opinion on that, but it's super irrelevant for ORM 3.

> should code be simplified with abstract classes? yes / no

We could do that. No necessary for ORM 3 support though, imho.

> should run separate tests with ORM 2.x and ORM 3.x? yes / no

As long as both versions are supported, we need to test against both. However, we can be lazy here: ORM 2 will be tested on the PHP 7 jobs automatically because ORM 3 requires PHP 8.1.

> If you look at `$parser->match()` methods, because of the `LexerType` enum, it can only be done with local ifs. Using abstract classes also reduces the footprint of this change.

If you just need to share some glue code among classes, you can use a trait instead. That's what traits are made for.

Reply:
> > If you look at `$parser->match()` methods, because of the `LexerType` enum, it can only be done with local ifs. Using abstract classes also reduces the footprint of this change.
> 
> If you just need to share some glue code among classes, you can use a trait instead. That's what traits are made for.

Interesting suggestion. I'll think about whether we could use traits to give a nicer solution to the `Lexer` contant vs `TokenType` enum/class problem. Where there is duplicate code, it's definitely a nicer solution, but the logic that only exists in one `FunctionNode` parser tends to make it harder to see and more work to throw ORM2 support in later.